### PR TITLE
Use different command types for each arity

### PR DESCRIPTION
**Pull request by jacereda**

### CODE REVIEW

**Comment by jacereda** [comment]
This should get rid of the ugly errors introduced to address incomplete patterns in the `Commands` module.

Be careful, there're some bits I'm not sure about. I'll start a review to ask about them.

File: src/Eval.hs [hunk 1]
```diff
@@ -361,8 +389,8 @@ eval ctx xobj@(XObj o info ty) preference =
         [XObj Address _ _, value] ->
           specialCommandAddress ctx value
         [] -> pure (ctx, dynamicNil)
-        _ -> do
-          pure (evalError ctx ("I did not understand the form `" ++ pretty xobj ++ "`") (xobjInfo xobj))
+        x -> error $ show x
+    -- _ -> pure (evalError ctx ("I did not understand the form `" ++ pretty xobj ++ "`") (xobjInfo (traceShowId xobj)))
```
Comment (jacereda) [inline src/Eval.hs:393]: I messed this.

File: src/Eval.hs [hunk 2]
```diff
@@ -315,13 +315,41 @@ eval ctx xobj@(XObj o info ty) preference =
               case res of
                 Right xobj' -> macroExpand ctx' xobj'
                 Left _ -> pure (ctx, res)
-        XObj (Lst [XObj (Command callback) _ _, _, _]) _ _ : args ->
+        [XObj (Lst [XObj (Command (NullaryCommandFunction nullary)) _ _, _, _]) _ _] ->
```
Comment (jacereda) [inline src/Eval.hs:318]: I guess I should pattern-match to catch wrong number of arguments in all of these.

File: src/Expand.hs [hunk 3]
```diff
@@ -201,8 +201,24 @@ expand eval ctx xobj =
               Right (XObj (Lst [XObj Macro _ _, _, XObj (Arr _) _ _, _]) _ _) ->
                 --trace ("Found macro: " ++ pretty xobj ++ " at " ++ prettyInfoFromXObj xobj)
                 eval ctx'' xobj
-              Right (XObj (Lst [XObj (Command callback) _ _, _]) _ _) ->
-                getCommand callback ctx args
+              Right (XObj (Lst [XObj (Command (NullaryCommandFunction nullary)) _ _, _, _]) _ _) ->
+                nullary ctx''
+              Right (XObj (Lst [XObj (Command (UnaryCommandFunction unary)) _ _, _, _]) _ _) ->
+                case expandedArgs of
```
Comment (jacereda) [inline src/Expand.hs:207]: Should these use `args` or `expandedArgs`?

**Comment by jacereda** [comment]
Primitives could follow a similar approach.

**Review by hellerve — APPROVED** [review]
I think this is a good quality of life improvement!

File: src/Eval.hs [hunk 4]
```diff
@@ -361,8 +391,10 @@ eval ctx xobj@(XObj o info ty) preference =
         [XObj Address _ _, value] ->
           specialCommandAddress ctx value
         [] -> pure (ctx, dynamicNil)
-        _ -> do
-          pure (evalError ctx ("I did not understand the form `" ++ pretty xobj ++ "`") (xobjInfo xobj))
+        _ -> pure (evalError ctx ("I did not understand the form `" ++ pretty xobj ++ "`") (xobjInfo xobj))
+    badArity params args i = case checkArity params args of
```
[the remark this answers is not in the document]
Comment (jacereda) [reply src/Eval.hs]: Done.

File: src/Expand.hs (hunk 3 above)
Comment (eriksvedang) [reply src/Expand.hs:207]: Not sure, do you think it ever matters? I'll merge as is, seems fine...

**Review by eriksvedang — APPROVED** [review]

**Pull request merged**